### PR TITLE
docs: add AutoQEC system design document

### PR DESCRIPTION
**Pull request by xiejingu**
Comprehensive design doc covering the MVP scope: Surface Code MLP decoder with MWPM + BP-OSD baselines, Stim data pipeline, configurable training, and HTML report generation via ECharts.

## Summary by Sourcery

Add a comprehensive system design document for the AutoQEC AI-driven quantum error correction decoder and supporting brainstorming notes, formalizing the planned architecture, pipelines, and experimental scope.

Documentation:
- Add a detailed design.md describing the AutoQEC MVP, including goals, data pipeline, model configuration, baselines, evaluation protocol, and system architecture.
- Add an AutoQEC_Brainstorm Jupyter notebook capturing exploratory ideas, risks, and phased roadmap for the AutoQEC project.

### CODE REVIEW

**Comment by qualit527** [comment]
关键发现

  HIGH（阻塞合并）

  1. independent_verify() 根本没评估 predecoder（autoqec/eval/independent_eval.py:79-112）
    - 在非-None 分支里 pred_errors = plain_errors.copy()，然后 ler_shuffled = ler_plain，最后 ler_holdout_ci=(plo, phi) 用的是 plain baseline 的
   bootstrap。
    - 结果：verifier 区分不出"真改进"和"啥都没做"；ablation check 是空的；holdout CI 报错对象。
    - 修复方向：真正 reload 模型 → 在同一批 holdout shots 上跑 plain + predecoder → 用 pred_errors 算 CI → shuffle 参数后再跑得到真正的
  ler_shuffled。
  2. _load_predecoder() 不认现 main 上 Runner 写的 checkpoint 格式（independent_eval.py:39-48）
    - Xie 只接受 {"model": <pickle>} 的 ckpt；Lin 的 Runner 写 {"class_name", "state_dict", "output_mode", "dsl_config"}。
    - 结果：每次 reload 都返回 None，落入 identity fallback，任何真实 decoder 都不可能进入 VERIFIED。
    - 修复方向：用 dsl_config + class_name 重建 module + load_state_dict；或让 verifier entry 接受整个 round_dir 复用 Runner 的 compile 路径。
  3. Memorizer cheater 是惰性的（autoqec/cheaters/memorize.py:11-30 + tests/test_reward_hacking.py:6-18）
    - MemorizerPredecoder.table 从未填充；save_memorizer_ckpt() 存空模型；测试只断言 verdict in ("SUSPICIOUS","FAILED")。
    - 结果：guard 即便完全坏掉，这条 cheater 测试也仍然过。Reward-hacking 探针没校准任何东西。
    - 修复方向：实现真的 train-seed memorize 步骤；加一条对照测试证明它在 holdout 隔离被绕过时获胜，在正确 guard 下失败。
  4. Seed 隔离没按契约强制（independent_eval.py:29-36, 58-75）
    - Verifier 接受调用方传入的 holdout_seeds 列表，只检查与 train/val 不重叠；不强制要求落在 env_spec.noise.seed_policy.holdout 范围内。
    - per_seed_shots = n_shots // len(holdout_seeds) 是 floor，名义 min_shots_verify=200_000 实际可能少跑。
    - 修复方向：要么 verifier 自己从 seed_policy.holdout 选 seed，要么校验调用方传入的 seed ⊆ holdout 范围；用 ceil 保证总 shots 不少于
  minimum。

  MEDIUM

  5. BP-OSD wrapper 与 main 上 backend_adapter 行为不一致（bposd_wrap.py:8-18）
    - 新 wrapper 硬编码 error_rate=... + osd_method="osd_e"/"osd_0"；现有 backend_adapter.py 用 osd_method="osd_cs" + channel_probs 支持
  soft_priors。
    - 项目里现在有两条 OSD 解码路径，语义不同。
    - 修复方向：复用现有 adapter，或抽出共享 helper。
  6. Pareto helper 无 tie 处理 / 不去重 / 不排序 / 不截断（pareto/front.py:4-23）
    - 与 docs/contracts/round_dir_layout.md 期望的"按 -delta_ler 排序、capped"矛盾。
  7. 测试都是 happy-path smoke，捕不到上面任何一条 bug
    - "诚实 predecoder" 测试在 verifier 完全忽略 ckpt 时仍然通过。
    - bootstrap 测试只测一个大样本二项；未测复现性、退化输入。
    - BP-OSD 测试接受 bit 0 或 bit 1，无法检测 OSD 方法/order 错误。

  MEDIUM（文档）

  8. design.md 与 main 上的冻结契约严重对不上（design.md:154-193, 303-449）
    - 写了 noise.error_rates / data.* / training.* / evaluation.* 字段，output mode 写 observable_flip / pauli_per_qubit / full_error ——
  实际权威契约是 EnvSpec.noise.p / constraints / baseline_decoders / classical_backend / eval_protocol，output mode 是 hard_flip / soft_priors。
    - 写了 autoqec/data, autoqec/models, autoqec/train, autoqec/viz, main.py, requirements.txt —— 现在树里是 autoqec/runner,
  autoqec/orchestration, autoqec/agents, autoqec/tools + cli/autoqec.py + pyproject.toml。
    - 整篇架构章节都已 stale。

  LOW

  9. design.md BP-OSD 章用 from ldpc import BpOsdDecoder，实际 ldpc 包是 bposd_decoder。
  10. README 第 84 行把 Xie plan 链接写成 docs/superpowers/pls/...，应是 plans/。

**Comment by qualit527** [comment]
本轮新发现

  HIGH / 潜伏 crash（必须修，影响任何真实 Runner ckpt）

  _shuffle_model_params 是坏的。autoqec/eval/independent_eval.py:193-199：

  param.data = param.data[
      torch.randperm(param.data.numel()).reshape(param.data.shape)
  ]

  对任何 shape[0] < numel 的参数（即所有 2D 权重如 nn.Linear.weight），生成的 index 值域是 [0, numel) 而非 [0, shape[0])，走 torch/numpy
  advanced indexing 时会 IndexError。我用同构的 numpy 语义验证过会崩。

  当前 33 passed 没暴露出来，纯因为唯一加载的非 None model 是 MemorizerPredecoder，它只有 self.dummy = Parameter(zeros(1))（1D
  size-1）；test_independent_verify_identity_predecoder 走 None 分支；其他测试都不 compile GNN。

  → 换成 flat = param.data.view(-1); param.data = flat[torch.randperm(flat.numel())].view(param.shape)，并补一条「加载
  compile_predecoder(gnn_small) + run verify」的集成测试。

  H2 — Memorizer 的 ablation 是空操作

  Memorizer 的「学到的东西」在 self.table（dict），ablation 却只 shuffle self.dummy（永远是零张量）。结果 pred_errors ==
  shuffled_errors，ablation_sanity_ok 自动 True，不承载任何信号。当前 Demo 4 之所以仍被判 FAILED，靠的是 delta < -ci_half（hard_flip=0 让 MWPM
  看到全零 syndrome → 几乎每发都错），跟 ablation 无关。若要 ablation 真正覆盖 Memorizer 这类样本，应该 ablate table 本身（清空 / 随机置换
  corrections）。

  MED

  - M1 Demo 5 run.sh 路径不对：diagnose runs/demo-5/round_0 会 rd.glob("round_*") 返回空 → 输出 No round dirs → 违反 README 自己写的 "CLI prints
   structured round stats"。应改为 diagnose runs/demo-5，或让 CLI 兼容 round_dir。
  - M2 cli verify 硬编码 holdout[:100]（cli/autoqec.py:129-131），与 --n-shots 解耦；200k shots × 100 seeds = 2000
  shots/seed，小环境下还行但没暴露给用户。
  - M3 torch.load(..., weights_only=False)（independent_eval.py:175）允许任意 pickle 反序列化；作为「独立审计面」的 verifier 应当首选
  weights_only=True 走 state_dict 分支，把 pickle model 走 legacy flag。
  - M4 n_shots = math.ceil(n_shots or …) 是 no-op（int→int）；真正的 ceil 在 _sample_holdout 里已经做了，删掉免误导。
  - M5 test_memorizer_fails_independent_verify 名实不符：传 artifacts=None 落入空 table 分支，实际测的是 identity-like memorizer。真正的「填表后
   holdout 失败」在第二条测试。
  - M6 Isolation 口径：lint 只看 independent_eval.py → autoqec.runner.*，当前文件确实干净 ✅；但它现在硬依赖
  autoqec.decoders.{backend_adapter,dsl_compiler,pymatching_wrap}。验证端与训练端共享 classical-backend 代码意味着 backend 的 bug
  在两端都藏得住——这是务实取舍，但应在注释里说清楚「isolation 指的是 runner 层面，不是 decoder 层面」。另外 memorize.py 引
  autoqec.runner.data.sample_syndromes（规则未禁，但把 runner 耦合渗进 cheater 了）。

  LOW

  - L1 Demo 4 walkthrough "memorizer has no learned weights" 字面不准（table 就是学到的），措辞问题。
  - L2 _select_seeds 在 ae42cbc 版本里已不被使用，死代码。

  建议

  合并前至少修 H1（否则 verifier 对真实 ckpt 直接 crash，整个 publishability gate 就瞎了）和 M1（demo 自相矛盾）。H2 / M5 建议在下一轮 Demo 4
  调优里一起处理，同时加一条集成测试：compile_predecoder(gnn_small) → torch.save(Runner format) → independent_verify，跑通 shuffle 路径再 green
  才算数。

**Comment by qualit527** [comment]
Findings

  [HIGH] Dead-link FileNotFoundError at runtime — docs/plans/2026-04-21-autoqec-person-a-chen.md:814
  The commit deletes docs/superpowers/specs/2026-04-20-autoqec-design.md but the chen plan still contains executable
  Path("docs/superpowers/specs/2026-04-20-autoqec-design.md").read_text(). Any agent/human running that step will crash.

  [HIGH] Spec deleted with 13 live references — README top-level Spec link, .claude/agents/autoqec-coder.md:12 (Coder subagent is told to read
  the deleted file for DSL schema), all 4 moved plans, 3 person briefs, 2 knowledge INDEX files. Deleting the repo's declared "source of truth"
  in a fix: commit without replacement or redirect is not acceptable.

  [HIGH] Doc reorg incomplete — docs/plans/2026-04-21-autoqec-master.md:19-21, 481
  Moved plan still points to the old docs/superpowers/plans/... path for the person plans, and self-references the pre-move location. Internal
  link rot.

  [MEDIUM] Silent revert of PR #3 (pinyin → Chinese) — all moved plan/spec titles+bodies. PR #3 (cecda83) romanized names explicitly "for CI";
  this commit undoes that across 7 files. Commit message doesn't mention it. Either preserve pinyin or explain why the CI rationale no longer
  applies.

  [MEDIUM] Commit message mis-scopes the change — subject says "lint + reorganize docs" but body hides (a) spec deletion, (b) romanization
  reversal, (c) brainstorm notebook deletion. Split into fix(lint) + docs(reorg) or expand the body.

  [LOW] Python lint edits themselves — clean. pymatching top-level import was genuinely unused (verifier uses PymatchingBaseline). total was
  dead. l→line is a pure rename. Removed pytest / Path in tests were unused with no side-effect chain.

**Comment by qualit527** [comment]
Findings

  [MED] design.md:~234 — wrong ldpc API surface
  Snippet uses from ldpc import BpOsdDecoder with parity_check_matrix=…, bp_method="minimum_sum", osd_method="osd_0". Actual ldpc API is from ldpc
  import bposd_decoder (lowercase, positional matrix, bp_method="ms"/"ps", osd_method="osd0"/"osd_cs") — confirmed by current repo usage in
  autoqec/decoders/backend_adapter.py:7. Anyone following the doc gets a non-working import.

  [MED] design.md:~217 — evaluation target under-specified for baselines
  MWPM matching.decode(...) and BP-OSD decoder.decode(...) return corrections in their own spaces, but §5.5 compares everything against
  observable_flip. Missing projection from correction → logical-observable prediction. Easy to end up comparing unlike quantities.

  [MED] design.md:~168, 324-326 — drift from later-frozen contracts, not flagged
  Doc uses noise.error_rates, data.*, training.*, evaluation.baselines, output modes observable_flip / pauli_per_qubit / full_error, main.py,
  requirements.txt. Frozen docs/contracts/interfaces.md uses EnvSpec.noise.p, constraints, baseline_decoders, classical_backend, eval_protocol, output
  modes hard_flip / soft_priors; repo uses pyproject.toml + cli/autoqec.py. Needs a "pre-Phase-0 historical draft" banner + a mapping table.

  [MED] design.md:~360 — "extract Hx, Hz from Stim circuit" is hand-waved
  Stim doesn't expose a first-class "give me CSS parity-check matrices for this circuit" op. The BP-OSD baseline's key dependency has no named
  converter. Either scope BP-OSD to envs where H is supplied explicitly, or name the concrete converter.

  [LOW] design.md:~546 — repro story weaker than it claims
  Only torch.manual_seed(42) + np.random.seed(42). Missing Python random, Stim sampler seeds, train/val/holdout partitioning. Contradicts the
  "one-command reproducibility" success metric and the later frozen seed-isolation contract.

**Comment by qualit527** [comment]
Findings

  [MED] ldpc API still drifting (docs/specs/…:~318)
  Doc now says the current impl uses bp_method="ms"/"ps" and osd_method="osd0"/"osd_cs". 实际 autoqec/decoders/backend_adapter.py:26,33 只传
  osd_method="osd_cs"，根本没 bp_method，也没出现过 "osd0"。Import 是修对了（from ldpc import bposd_decoder），但"current
  implementation"这句话仍然错。Fix: 要么把这段改成与 backend_adapter.py 一致的真实调用，要么改成"see repo for concrete API"。

  [MED] LER 公式对 Stim 不成立 (docs/specs/…:~435)
  Doc: (correction[:, :n_obs] != observables).any(axis=1). 实际 runner._failure_rate() 是 (predictions != targets).any(axis=1).mean()，没有 [:, :n_obs]
   切片。decode_with_predecoder 对 MWPM 路径已经返回 observable-shape 的 predictions，不需要切；切片语义只在完整 correction
  向量时才有意义，与代码不匹配。应该写 predicted_observables != observables。

  [MED] CLI drift 只修了 Makefile，spec 里还有 5 处 autoqec run/verify/pareto（行 323, 464, 479, 481, 646, 669）
  pyproject.toml 没有 [project.scripts] 条目，bare autoqec 命令不存在。要么全部改成 python -m cli.autoqec …，要么在 pyproject.toml 加个 console-script
  让 bare 命令真的能跑。

**Review by qualit527 — APPROVED** [review]
